plugins/newsdownloader.koplugin/feed_source.lua

```diff
@@ -0,0 +1,364 @@
+--                    total_items = (limit == 0)
+--                        and #document.rss.channel.item
+--                        or limit
+                end,
+                function()
+                    -- Atom callback
+                    feed_title = FeedSource:getFeedTitle(document.feed.title)
+                    feed_items = document.feed.entry
+                    total_items = #document.feed.entry
+                end
+            )
+    end)
+    if ok then
+        document.title = feed_title
+        document.items = feed_items
+        document.total_items = total_items
+        return document
+    else
+        error("Could not initialize feed document")
+    end
+end
+--
+function FeedSource:getFeedItems(feed, progress_callback, error_callback)
+    local limit = tonumber(feed.config.limit)
+    local download_full_article = feed.config.download_full_article ~= false
+    local total_items = (limit == 0)
+        and feed.document.total_items
+        or limit
+
+    local feed_html = {}
+    -- Download each ite0m in the feed
+    for index, item in pairs(feed.document.items) do
+        -- If limit has been met, stop downloading feed.
+        if limit ~= 0 and index - 1 == limit then
+            break
+        end
+        -- Display feedback to user.
+        progress_callback(T(
+            _("Getting item %1 of %2 from %3"),
+            index,
+            total_items,
+            feed.document.title
+        ))
+        -- Download the article's HTML.
+        local ok, response = pcall(function()
+                return self:getItemHtml(
+                    item
+                )
+        end)
+        -- Add the result to our table, or send a
+        -- result to the error callback.
+        if ok then
+            table.insert(
+                feed_html,
+                response
+            )
+        else
+            error_callback(
+                T(_("Could not get feed item for: %1"),
+                  feed.document.title
+                )
+            )
+        end
+
+    end
+
+    if #feed_html > 0 then
+        return feed_html
+    else
+        return nil
+    end
+end
+
+   -- self:outputEpub(feed_html[0], feed_output_dir, '')
+--    self:outputEpub(html, feed_output_dir, article_message)
+--
+
+function FeedSource:getFeedType(document, rss_cb, atom_cb)
+    -- Check to see if the feed uses RSS.
+    local is_rss = document.rss
+        and document.rss.channel
+        and document.rss.channel.title
+        and document.rss.channel.item
+        and document.rss.channel.item[1]
+        and document.rss.channel.item[1].title
+        and document.rss.channel.item[1].link
+    -- Check to see if the feed uses Atom.
+    local is_atom = document.feed
+        and document.feed.title
+        and document.feed.entry[1]
+        and document.feed.entry[1].title
+        and document.feed.entry[1].link
+    -- Setup the feed values based on feed type
```

Comment:
(For this kind of construct, the fact that you choose the (correct, 4 spaces :) block level indentation makes it confusing - one would quick read it as it's a block. You're allowed to add a lot more indentation to avoid this, and to get some kind of alignment on the right to make this prettier and easier to read, like the `and` with the first `document`, or all the `document` with each others.)

Reply:
Thanks, that's helpful and I will revise accordingly :)